File: extras/Python/cmd_messenger/connection_manager.py
"""``ConnectionManager`` — port of C# ``ConnectionManager``.

Abstract base for transport-specific connection managers. Owns an
:class:`AsyncWorker` that switches between scan / connect / watchdog modes
and fires :data:`connection_found` / :data:`connection_timeout` events.
"""
from __future__ import annotations

import enum
import threading
import time
from abc import ABC, abstractmethod
from typing import Optional

from . import time_utils
from .async_worker import AsyncWorker
from .cmd_messenger import CmdMessenger
from .enums import ReceiveQueue, SendQueue, UseQueue
from .event import Event
from .received_command import ReceivedCommand
from .send_command import SendCommand


class Mode(enum.Enum):
    WAIT = "wait"
    CONNECT = "connect"
    SCAN = "scan"
    WATCHDOG = "watchdog"


class DeviceStatus(enum.Enum):
    NOT_AVAILABLE = "not_available"
    AVAILABLE = "available"
    IDENTITY_MISMATCH = "identity_mismatch"


class ConnectionManagerProgressEventArgs:
    def __init__(self, level: int, description: str) -> None:
        self.level = level
        self.description = description

    def __repr__(self) -> str:
        return f"<Progress level={self.level} {self.description!r}>"


class ConnectionManager(ABC):
    """Abstract connection manager."""

    def __init__(
        self,
        cmd_messenger: CmdMessenger,
        identify_command_id: int = 0,
        unique_device_id: Optional[str] = None,
    ) -> None:
        if cmd_messenger is None:
            raise ValueError("Command Messenger is null.")

        #: Fires () when the device fails to respond.
        self.connection_timeout: Event = Event()
        #: Fires () when a device responds successfully.
        self.connection_found: Event = Event()
        #: Fires (ConnectionManagerProgressEventArgs) for log messages.
        self.progress: Event = Event()

        self._cmd_messenger = cmd_messenger
        self._identify_command_id = identify_command_id
        self._unique_device_id = unique_device_id

        self.watchdog_timeout: int = 3000
        self.watchdog_retry_timeout: int = 1500
        self.watchdog_tries: int = 3
        self._watchdog_enabled: bool = False

        #: Whether to persist last-good settings (subclass-specific).
        self.persistent_settings: bool = False
        #: Whether to scan for devices vs connect to a known device.
        self.device_scan_enabled: bool = True

        #: Currently connected to a device.
        self.connected: bool = False

        self._mode: Mode = Mode.WAIT
        self._last_check_time: int = 0
        self._next_time_out_check: int = 0
        self._watchdog_tries_used: int = 0

        self._worker = AsyncWorker(self._do_work, name="ConnectionManager")

        if unique_device_id:
            cmd_messenger.attach(identify_command_id, self._on_identify_response)

    # ------------------------------------------------------------------
    # Watchdog property
    # ------------------------------------------------------------------
    @property
    def watchdog_enabled(self) -> bool:
        return self._watchdog_enabled

    @watchdog_enabled.setter
    def watchdog_enabled(self, value: bool) -> None:
        if value and not self._unique_device_id:
            raise RuntimeError("Watchdog can't be enabled without Unique Device ID.")
        self._watchdog_enabled = value

    @property
    def mode(self) -> Mode:
        return self._mode

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------
    def start_connection_manager(self) -> None:
        if not self._worker.is_running:
            self._worker.start()
        if self.device_scan_enabled:
            self._start_scan()
        else:
            self._start_connect()

    def stop_connection_manager(self) -> None:
        if self._worker.is_running:
            self._worker.stop()
        self._disconnect()

    def dispose(self) -> None:
        self.stop_connection_manager()

    def __enter__(self) -> "ConnectionManager":
        return self

    def __exit__(self, *exc_info) -> None:
        self.dispose()

    # ------------------------------------------------------------------
    # Events
    # ------------------------------------------------------------------
    def _connection_found_event(self) -> None:
        self._mode = Mode.WAIT
        if self._watchdog_enabled:
            self._start_watchdog()
        try:
            self.connection_found()
        except Exception:
            pass

    def _connection_timeout_event(self) -> None:
        self._mode = Mode.WAIT
        self._disconnect()
        try:
            self.connection_timeout()
        except Exception:
            pass
        if self._watchdog_enabled:
            self._stop_watchdog()
            if self.device_scan_enabled:
                self._start_scan()
            else:
                self._start_connect()

    def _log(self, level: int, message: str) -> None:
        try:
            self.progress(ConnectionManagerProgressEventArgs(level, message))
        except Exception:
            pass

    # ------------------------------------------------------------------
    # Identify / watchdog
    # ------------------------------------------------------------------
    def _on_identify_response(self, response_command: ReceivedCommand) -> None:
        if response_command.ok and self._unique_device_id:
            self._validate_device_unique_id(response_command)

    def _arduino_available(self, timeout_ms: int, tries: int = 1) -> DeviceStatus:
        for i in range(1, tries + 1):
            if tries > 1:
                self._log(3, f"Polling Arduino, try # {i}")
            challenge = SendCommand(
                self._identify_command_id,
                self._identify_command_id,
                ack_cmd_id=self._identify_command_id,
                timeout=timeout_ms,
            )
            response = self._cmd_messenger.send_command(
                challenge,
                SendQueue.IN_FRONT_QUEUE,
                ReceiveQueue.DEFAULT,
                UseQueue.BYPASS_QUEUE,
            )
            if response.ok and self._unique_device_id:
                return (
                    DeviceStatus.AVAILABLE
                    if self._validate_device_unique_id(response)
                    else DeviceStatus.IDENTITY_MISMATCH
                )
            if response.ok:
                return DeviceStatus.AVAILABLE
        return DeviceStatus.NOT_AVAILABLE

    def _validate_device_unique_id(self, response_command: ReceivedCommand) -> bool:
        valid = self._unique_device_id == response_command.read_string_arg()
        if not valid:
            self._log(3, "Invalid device response. Device ID mismatch.")
        return valid

    # ------------------------------------------------------------------
    # Worker job
    # ------------------------------------------------------------------
    def _do_work(self) -> bool:
        if self._mode == Mode.SCAN:
            self._do_work_scan()
        elif self._mode == Mode.CONNECT:
            self._do_work_connect()
        elif self._mode == Mode.WATCHDOG:
            self._do_work_watchdog()
        time.sleep(0.1)
        return True

    @abstractmethod
    def _do_work_connect(self) -> None:
        """Subclass: try to connect using current settings."""

    @abstractmethod
    def _do_work_scan(self) -> None:
        """Subclass: scan available devices for a match."""

    def _do_work_watchdog(self) -> None:
        last_line_time_stamp = self._cmd_messenger.last_received_command_time_stamp
        current = time_utils.millis()
        if current < self._next_time_out_check:
            return
        if last_line_time_stamp >= self._last_check_time:
            self._log(3, "Successful watchdog response.")
            self._last_check_time = current
            self._next_time_out_check = self._last_check_time + self.watchdog_timeout
            self._watchdog_tries_used = 0
            return

        if self._watchdog_tries_used >= self.watchdog_tries:
            self._log(2, f"Watchdog received no response after final try #{self.watchdog_tries}")
            self._watchdog_tries_used = 0
            self._mode = Mode.WAIT
            self._connection_timeout_event()
            return

        self._cmd_messenger.send_command(SendCommand(self._identify_command_id))
        self._watchdog_tries_used += 1
        self._last_check_time = current
        self._next_time_out_check = self._last_check_time + self.watchdog_retry_timeout
        if self._watchdog_tries_used == 1:
            self._log(
                3,
                f"Watchdog detected no communication for {self.watchdog_timeout / 1000.0}s, asking for response",
            )
        else:
            self._log(3, f"Watchdog received no response, performing try #{self._watchdog_tries_used}")

    # ------------------------------------------------------------------
    # State transitions
    # ------------------------------------------------------------------
    def _disconnect(self) -> bool:
        if self.connected:
            self.connected = False
            return self._cmd_messenger.disconnect()
        return True

    def _start_watchdog(self) -> None:
        if self._mode != Mode.WATCHDOG and self.connected:
            self._log(1, "Starting Watchdog.")
            self._last_check_time = time_utils.millis()
            self._next_time_out_check = self._last_check_time + self.watchdog_timeout
            self._watchdog_tries_used = 0
            self._mode = Mode.WATCHDOG

    def _stop_watchdog(self) -> None:
        if self._mode == Mode.WATCHDOG:
            self._log(1, "Stopping Watchdog.")
            self._mode = Mode.WAIT

    def _start_scan(self) -> None:
        if self._mode != Mode.SCAN and not self.connected:
            self._log(1, "Starting device scan.")
            self._mode = Mode.SCAN

    def _stop_scan(self) -> None:
        if self._mode == Mode.SCAN:
            self._log(1, "Stopping device scan.")
            self._mode = Mode.WAIT

    def _start_connect(self) -> None:
        if self._mode != Mode.CONNECT and not self.connected:
            self._log(1, "Start connecting to device.")
            self._mode = Mode.CONNECT

    def _stop_connect(self) -> None:
        if self._mode == Mode.CONNECT:
            self._log(1, "Stop connecting to device.")
            self._mode = Mode.WAIT

    # ------------------------------------------------------------------
    # Persistence (subclass-specific)
    # ------------------------------------------------------------------
    def _store_settings(self) -> None:  # pragma: no cover - hook
        pass

    def _read_settings(self) -> None:  # pragma: no cover - hook
        pass

File: extras/TypeScript/src/connectionManager.ts
import { CmdMessenger } from './cmdMessenger.js';
import { ReceiveQueue, SendQueue, UseQueue } from './enums.js';
import { ReceivedCommand } from './receivedCommand.js';
import { SendCommand } from './sendCommand.js';
import { delay, millis } from './timeUtils.js';
import { TypedEmitter } from './typedEmitter.js';

export enum ConnectionManagerMode {
  Wait = 'wait',
  Connect = 'connect',
  Scan = 'scan',
  Watchdog = 'watchdog'
}

export enum DeviceStatus {
  NotAvailable = 'not_available',
  Available = 'available',
  IdentityMismatch = 'identity_mismatch'
}

export interface ConnectionManagerProgress {
  level: number;
  description: string;
}

interface ConnectionManagerEvents {
  connectionFound: () => void;
  connectionTimeout: () => void;
  progress: (progress: ConnectionManagerProgress) => void;
}

export abstract class ConnectionManager {
  readonly events = new TypedEmitter<ConnectionManagerEvents>();

  connected = false;
  watchdogTimeout = 3000;
  watchdogRetryTimeout = 1500;
  watchdogTries = 3;
  persistentSettings = false;
  deviceScanEnabled = true;

  private running = false;
  private modeValue = ConnectionManagerMode.Wait;
  private watchdogEnabledValue = false;
  private lastCheckTime = 0;
  private nextTimeoutCheck = 0;
  private watchdogTriesUsed = 0;

  protected constructor(
    protected readonly cmdMessenger: CmdMessenger,
    private readonly identifyCommandId = 0,
    private readonly uniqueDeviceId?: string
  ) {
    if (this.uniqueDeviceId !== undefined && this.uniqueDeviceId.length > 0) {
      this.cmdMessenger.attach(this.identifyCommandId, (command) => this.onIdentifyResponse(command));
    }
  }

  get mode(): ConnectionManagerMode {
    return this.modeValue;
  }

  get watchdogEnabled(): boolean {
    return this.watchdogEnabledValue;
  }

  set watchdogEnabled(value: boolean) {
    if (value && (this.uniqueDeviceId === undefined || this.uniqueDeviceId.length === 0)) {
      throw new Error("Watchdog can't be enabled without Unique Device ID.");
    }
    this.watchdogEnabledValue = value;
  }

  startConnectionManager(): void {
    if (!this.running) {
      this.running = true;
      void this.runLoop();
    }

    if (this.deviceScanEnabled) {
      this.startScan();
    } else {
      this.startConnect();
    }
  }

  stopConnectionManager(): void {
    this.running = false;
    this.modeValue = ConnectionManagerMode.Wait;
    void this.disconnect();
  }

  dispose(): void {
    this.stopConnectionManager();
    this.events.clear();
  }

  protected abstract doWorkConnect(): Promise<void>;
  protected abstract doWorkScan(): Promise<void>;

  protected async arduinoAvailable(timeoutMs: number, tries = 1): Promise<DeviceStatus> {
    for (let i = 1; i <= tries; i += 1) {
      if (tries > 1) {
        this.log(3, `Polling Arduino, try # ${i}`);
      }

      const challenge = new SendCommand(this.identifyCommandId).withAck(this.identifyCommandId, timeoutMs);
      const response = await this.cmdMessenger.sendCommand(
        challenge,
        SendQueue.InFrontQueue,
        ReceiveQueue.Default,
        UseQueue.BypassQueue
      );

      if (response.ok && this.uniqueDeviceId !== undefined && this.uniqueDeviceId.length > 0) {
        return this.validateDeviceUniqueId(response) ? DeviceStatus.Available : DeviceStatus.IdentityMismatch;
      }

      if (response.ok) {
        return DeviceStatus.Available;
      }
    }

    return DeviceStatus.NotAvailable;
  }

  protected connectionFoundEvent(): void {
    this.modeValue = ConnectionManagerMode.Wait;
    if (this.watchdogEnabledValue) {
      this.startWatchdog();
    }
    this.events.emit('connectionFound');
  }

  protected async connectionTimeoutEvent(): Promise<void> {
    this.modeValue = ConnectionManagerMode.Wait;
    await this.disconnect();
    this.events.emit('connectionTimeout');

    if (this.watchdogEnabledValue) {
      this.stopWatchdog();
      if (this.deviceScanEnabled) {
        this.startScan();
      } else {
        this.startConnect();
      }
    }
  }

  protected log(level: number, description: string): void {
    this.events.emit('progress', { level, description });
  }

  protected startScan(): void {
    if (this.modeValue !== ConnectionManagerMode.Scan && !this.connected) {
      this.log(1, 'Starting device scan.');
      this.modeValue = ConnectionManagerMode.Scan;
    }
  }

  protected stopScan(): void {
    if (this.modeValue === ConnectionManagerMode.Scan) {
      this.log(1, 'Stopping device scan.');
      this.modeValue = ConnectionManagerMode.Wait;
    }
  }

  protected startConnect(): void {
    if (this.modeValue !== ConnectionManagerMode.Connect && !this.connected) {
      this.log(1, 'Start connecting to device.');
      this.modeValue = ConnectionManagerMode.Connect;
    }
  }

  protected stopConnect(): void {
    if (this.modeValue === ConnectionManagerMode.Connect) {
      this.log(1, 'Stop connecting to device.');
      this.modeValue = ConnectionManagerMode.Wait;
    }
  }

  private async runLoop(): Promise<void> {
    while (this.running) {
      try {
        switch (this.modeValue) {
          case ConnectionManagerMode.Connect:
            await this.doWorkConnect();
            break;
          case ConnectionManagerMode.Scan:
            await this.doWorkScan();
            break;
          case ConnectionManagerMode.Watchdog:
            await this.doWorkWatchdog();
            break;
          case ConnectionManagerMode.Wait:
            break;
        }
      } catch (error) {
        this.log(2, error instanceof Error ? error.message : String(error));
      }
      await delay(100);
    }
  }

  private onIdentifyResponse(responseCommand: ReceivedCommand): void {
    if (responseCommand.ok && this.uniqueDeviceId !== undefined && this.uniqueDeviceId.length > 0) {
      this.validateDeviceUniqueId(responseCommand);
    }
  }

  private validateDeviceUniqueId(responseCommand: ReceivedCommand): boolean {
    const valid = this.uniqueDeviceId === responseCommand.readStringArg();
    if (!valid) {
      this.log(3, 'Invalid device response. Device ID mismatch.');
    }
    return valid;
  }

  private async doWorkWatchdog(): Promise<void> {
    const lastLineTimeStamp = this.cmdMessenger.lastReceivedCommandTimeStamp;
    const current = millis();

    if (current < this.nextTimeoutCheck) {
      return;
    }

    if (lastLineTimeStamp >= this.lastCheckTime) {
      this.log(3, 'Successful watchdog response.');
      this.lastCheckTime = current;
      this.nextTimeoutCheck = this.lastCheckTime + this.watchdogTimeout;
      this.watchdogTriesUsed = 0;
      return;
    }

    if (this.watchdogTriesUsed >= this.watchdogTries) {
      this.log(2, `Watchdog received no response after final try #${this.watchdogTries}`);
      this.watchdogTriesUsed = 0;
      this.modeValue = ConnectionManagerMode.Wait;
      await this.connectionTimeoutEvent();
      return;
    }

    await this.cmdMessenger.sendCommand(new SendCommand(this.identifyCommandId));
    this.watchdogTriesUsed += 1;
    this.lastCheckTime = current;
    this.nextTimeoutCheck = this.lastCheckTime + this.watchdogRetryTimeout;
    this.log(
      3,
      this.watchdogTriesUsed === 1
        ? `Watchdog detected no communication for ${this.watchdogTimeout / 1000}s, asking for response`
        : `Watchdog received no response, performing try #${this.watchdogTriesUsed}`
    );
  }

  private startWatchdog(): void {
    if (this.modeValue !== ConnectionManagerMode.Watchdog && this.connected) {
      this.log(1, 'Starting Watchdog.');
      this.lastCheckTime = millis();
      this.nextTimeoutCheck = this.lastCheckTime + this.watchdogTimeout;
      this.watchdogTriesUsed = 0;
      this.modeValue = ConnectionManagerMode.Watchdog;
    }
  }

  private stopWatchdog(): void {
    if (this.modeValue === ConnectionManagerMode.Watchdog) {
      this.log(1, 'Stopping Watchdog.');
      this.modeValue = ConnectionManagerMode.Wait;
    }
  }

  private async disconnect(): Promise<boolean> {
    if (!this.connected) {
      return true;
    }
    this.connected = false;
    return this.cmdMessenger.disconnect();
  }
}
